force-app/main/service/DonationHistoryService.cls

```diff
@@ -0,0 +1,176 @@
+
+        return donationHistory;
+    }
+    
+    /** 
+    * @author Salesforce.org
+    * @date 2021
+    * @description Verify hard credit condition.
+    */
+    private static List<DonationHistoryController.DonationHistory> verifyHardCreditCondition(List<OpportunityContactRole> donations){
+        List<DonationHistoryController.DonationHistory> hardCreditList = new List<DonationHistoryController.DonationHistory>();
+        for(OpportunityContactRole ocr: donations){
+            if (ocr.isPrimary) {
+                DonationHistoryController.DonationHistory hardCredit = new DonationHistoryController.DonationHistory(ocr, HARD_CREDIT);
+                hardCreditList.add(hardCredit);
+            }
+        }
+        return hardCreditList;
+    }
+    
+    /** 
+    * @author Salesforce.org
+    * @date 2021
+    * @description Verify soft credit condition.
+    */
+    private static List<DonationHistoryController.DonationHistory> verifySoftCreditCondition(List<OpportunityContactRole> donations){
+        //with the list of donations verify if the Role match the filter rule value on the NPSP_ClosedWon_Opps_SC filter group
+        Filter_Rule__mdt roles = [SELECT Constant__c FROM Filter_Rule__mdt WHERE QualifiedApiName = 'NPSP_OppsSoftCreditRoles'];
+        List<String> softCreditRoles = roles.Constant__c.split(';');
+        List<DonationHistoryController.DonationHistory> softCreditList = new List<DonationHistoryController.DonationHistory>();
+        for(OpportunityContactRole ocr: donations){
+            if (softCreditRoles.contains(ocr.Role) && ocr.isPrimary==false) {
+                DonationHistoryController.DonationHistory softCredit = new DonationHistoryController.DonationHistory(ocr, SOFT_CREDIT);
+                softCreditList.add(softCredit);
+            }
+        }
+        return softCreditList;
+    }
+
+    /** 
+    * @author Salesforce.org
+    * @date 2021
+    * @description Filter for ids already classified as Soft Credit.
+    */
+    private static List<Id> filterVerifiedIds(List<DonationHistoryController.DonationHistory> softCredit){
+        List<Id> donationsId = new List<Id>();
+        for(DonationHistoryController.DonationHistory dh: softCredit){
+            donationsId.add(dh.id);
+        }
+        return donationsId;
+    }
+
+    /** 
+    * @author Salesforce.org
+    * @date 2021
+    * @description Verify partial soft credit condition.
+    */
+    private static List<DonationHistoryController.DonationHistory> verifyPartialSoftCreditCondition(List<Id> donationsId, List<OpportunityContactRole> donations){
+        //search on PartialSoftCredit object for ids filtered before usin IN()
+        //only ids with not type of credit defined before
+        List<Id> partialSoftCreditIds = new List<Id>();
+        List<Partial_Soft_Credit__c> partialSoftCredit = [SELECT id, Amount__c, Contact_Role_ID__c FROM Partial_Soft_Credit__c WHERE Contact_Role_ID__c IN :donationsId];
```

Comment:
Let's take this query, and the one for OpportunityContactRole, and break them out into a DonationHistorySelector class as instance methods. We should also add `WITH SECURITY_ENFORCED` to those queries. There's a few more details in our slack channel on how this would work, and how we can mock data from that selector class to greatly speed up unit test performance.